Add package s3scanner

lavafroth opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

This PR adds a new package, s3scanner.
S3scanner is a security tool written in Go to scan for misconfigured S3 buckets across S3-compatible APIs.
Project repository: https://github.com/sa7mon/S3Scanner

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

lavafroth committed:
maintainers: add lavafroth

GetPsyched commented:
You are still not following the commit guidelines. Please give them a thorough read as mentioned in the previous PR #278854.

Also, please learn how to [rebase](https://git-scm.com/book/en/v2/Git-Branching-Rebasing), you do not need to create new PRs every time you change your commits.

GetPsyched commented on pkgs/top-level/all-packages.nix:
```diff
@@ -12964,6 +12964,10 @@ with pkgs;
     inherit (darwin.apple_sdk.frameworks) Security;
   };
 
+  s3scanner = callPackage ../by-name/s3/s3scanner/package.nix {
+    buildGoModule = buildGo119Module;
+  };
+
```
You do not need to do this when adding packages under the `by-name` convention
```suggestion
```

GetPsyched commented on pkgs/by-name/s3/s3scanner/package.nix:
```diff
@@ -0,0 +1,29 @@
+    license = licenses.mit;
+    maintainers = with maintainers; [ lavafroth ];
+    mainProgram = "s3scanner";
+    platforms = [ "x86_64-linux" "aarch64-linux" "x86_64-darwin" "aarch64-darwin" ];
```
This might be build-able on Windows and other operating systems as well. Consider one of the predefined defaults over at [lib/systems/plaforms.nix](https://github.com/NixOS/nixpkgs/blob/master/lib/systems/platforms.nix)

lavafroth committed:
s3scanner: init at 3.0.4

S3scanner is a security tool written in Go to scan for misconfigured S3 buckets across S3-compatible APIs.
Project repository: https://github.com/sa7mon/S3Scanner

GetPsyched commented:
Result of `nixpkgs-review pr 278867` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>s3scanner</li>
  </ul>
</details>

GetPsyched reviewed (approved):
##### Reviewed points

- [x] package path fits guidelines
- [x] package name fits guidelines
- [x] package version fits guidelines
- [x] package builds on x86_64-linux
- [x] executables tested on x86_64-linux
- [x] `meta.description` is set and fits guidelines
- [x] `meta.license` fits upstream license
- [x] `meta.platforms` is set
- [x] `meta.maintainers` is set
- [x] source is fetched using the appropriate function
- [x] the list of `phases` is not overridden

##### Possible improvements

Adding [passthru.tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)

#### Comments

Wondering why `doCheck` is set to false 🤔

lavafroth commented:
I set checks to false because IIRC almost all of them require network connection.

AndersonTorres commented on pkgs/by-name/s3/s3scanner/package.nix:
```diff
@@ -0,0 +1,29 @@
+    license = licenses.mit;
+    maintainers = with maintainers; [ lavafroth ];
+    mainProgram = "s3scanner";
+    platforms = platforms.all;
```
```suggestion
```

Because Go does not run in all packages.

AndersonTorres commented on pkgs/by-name/s3/s3scanner/package.nix:
```diff
@@ -0,0 +1,29 @@
+
+  vendorHash = "sha256-3Y1izt6xLg7aNJNqIEXROxR3IGAIIeptHlnoYEcuLew=";
+
+  doCheck = false;
```
```suggestion
  # Requires networking
  doCheck = false;
```

GetPsyched commented:
Removing `meta.platforms` wasn't the solution, but okay.

AndersonTorres commented on pkgs/by-name/s3/s3scanner/package.nix:
```diff
@@ -0,0 +1,30 @@
+  vendorHash = "sha256-3Y1izt6xLg7aNJNqIEXROxR3IGAIIeptHlnoYEcuLew=";
+
+  # Requires networking
+
```
```suggestion
```

AndersonTorres reviewed (approved):
Result of `nixpkgs-review pr 278867` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>s3scanner</li>
  </ul>
</details>

Besides some nits, LGTM

GetPsyched reviewed (approved):
LGTM, thanks for packaging this!

lavafroth commented:
Thanks for the review!
